### PR TITLE
[NNC] Fix loopnest.cache_accesses for reduce ops (fixed #59002)

### LINKED ISSUE #59002: [NNC] loopnest.cache_accesses transforms reduce op stmt wrongly
## 🐛 Bug

loopnest.cache_accesses performs a wrong transformation on the reduce op. 
For example, for the following stmt,
```
for (int x = 0; x < 32; x++) {
    for (int y = 0; y < 32; y++) {
      Z[x, y] = 0.f;
      for (int k = 0; k < 32; k++) {
        Z[x, y] = ReduceOp((Z[x, y]) + (X[x, k]) * (Y[k, y]), reduce_args={k});
```
`cache_accesses` `Z` buffer on `y` axis transforms the stmt into,
```
  for (int x = 0; x < 32; x++) {
    for (int i = 0; i < 1; i++) {
      for (int j = 0; j < 32; j++) {
        Z_local[i, j] = 0.f;
      }
    }
    for (int y = 0; y < 32; y++) {
      Z_local[0, y] = 0.f;
      for (int k = 0; k < 32; k++) {
        Z_local[0, y] = ReduceOp((Z_local[0, y]) + (X[x, k]) * (Y[k, y]), reduce_args={k});
      }
    }
    for (int i = 0; i < 1; i++) {
      for (int j = 0; j < 32; j++) {
        Z[i + x, j + 0] = ReduceOp((Z[i + x, j + 0]) + (Z_local[i, j]), reduce_args={});
      }
    }
  }
```
The transformed stmt does not equal to the original stmt, and produces different results.
## To Reproduce
Below is a test to reproduce the issue. The results of the transformed stmt are different from what the user expects, and commenting out the `cache_accesses` transformation on Line 31 yields correct results.

```
  1 import torch                                                                                                                                                                      
  2 
  3 class kernel_arena_scope(object):
  4     def __enter__(self):
  5         self.scope = torch._C._te.KernelScope()
  6 
  7     def __exit__(self, typ, val, traceback):
  8         self.scope = None
  9 
 10 with kernel_arena_scope():
 11     f32 = torch._C._te.Dtype.Float
 12     i32 = torch._C._te.Dtype.Int
 13     M = torch._C._te.ExprHandle.int(32)
 14     N = torch._C._te.ExprHandle.int(32)
 15     K = torch._C._te.ExprHandle.int(32)
 16 
 17     X = torch._C._te.Placeholder('X', f32, [M, K])
 18     Y = torch._C._te.Placeholder('Y', f32, [K, N])
 19 
 20     dim_args = [torch._C._te.DimArg(*args) for args in [(M, 'x'), (N, 'y')]]
 21 
 22     def compute(dims):
 23         m, n, k = dims
 24         return X.load([m, k]) * Y.load([k, n])
 25     Z = torch._C._te.Reduce('Z', dim_args, torch._C._te.Sum(), compute, [torch._C._te.DimArg(K, 'k')])
 26     loopnest = torch._C._te.LoopNest([Z])
 27     loopnest.simplify()
 28     print("original", loopnest)
 29 
 30     loops = loopnest.get_loops_for(Z)
 31     loopnest.cache_accesses(Z.buf(), 'Z_local', loops[1])
 32     print("after cache_accesses", loopnest)
 33     
 34     loopnest.flatten(loops)
 35     loopnest.prepare_for_codegen()
 36     stmt = torch._C._te.simplify(loopnest.root_stmt())
 37     codegen = torch._C._te.construct_codegen('llvm', stmt, [torch._C._te.BufferArg(x) for x in [X, Y, Z]])
 38     
 39     A = torch.rand(32, 32)
 40     B = torch.rand(32, 32)
 41     ref= torch.matmul(A, B)
 42     out = torch.rand(32, 32)
 43     codegen.call([A, B, out])
 44     torch.testing.assert_allclose(out, ref)
```
## Additional context

cc @gmagogsfm

### PR DESCRIPTION
**Pull request by huiguoo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59136 [NNC] Fix loopnest.cache_accesses for reduce ops (fixed #59002)**

Differential Revision: [D28768598](https://our.internmc.facebook.com/intern/diff/D28768598)

### CODE REVIEW

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28768598).

File: test/cpp/tensorexpr/test_reductions.cpp [hunk 1]
```diff
@@ -1331,34 +1331,138 @@ TEST(Reductions, ReductionCacheAccessesOuter) {
...
+#CHECK: for (int l1
+#CHECK:   d_local[l1] = 0.f
+#CHECK:   for (int n1
+#CHECK:     for (int m1
+#CHECK:       d_local[l1] = (d_local[l1]) + (scale[
+#CHECK:     }
+#CHECK:   }
+#CHECK: }
+#CHECK: for (int i
+#CHECK:   sum[i] = d_local[i]
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/test_reductions.cpp:1359]: Just for my understanding: this is the case that was incorrect previously, right? Were we generating a reduceOp here?

**Review by ZolotukhinM — APPROVED** [review]
Looks good to me!

File: test/cpp/tensorexpr/test_reductions.cpp (hunk 1 above)
Comment (huiguoo) [reply test/cpp/tensorexpr/test_reductions.cpp:1359]: Yes, that's the case we hit before: the `cache_accesses` is performed on the operator axis and we still got a reduceOp which we shouldn't.

**Comment by huiguoo** [comment]
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28768598).
